File: objective-c/test/Ice/operations/BatchOnewaysAMI.m
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#import <objc/Ice.h>
#import <TestCommon.h>
#import <OperationsTest.h>

#import <Foundation/Foundation.h>

void
batchOnewaysAMI(id<TestOperationsMyClassPrx> p)
{
    ICEByte buf1[10 * 1024];
    TestOperationsMutableByteS *bs1 = [TestOperationsMutableByteS dataWithBytes:buf1 length:sizeof(buf1)];

    id<TestOperationsMyClassPrx> batch = [TestOperationsMyClassPrx uncheckedCast:[p ice_batchOneway]];

    [batch end_ice_flushBatchRequests:[batch begin_ice_flushBatchRequests]]; // Empty flush
    [[batch begin_ice_flushBatchRequests] isCompleted]; // Empty flush
    [[batch begin_ice_flushBatchRequests] isSent]; // Empty flush
    [[batch begin_ice_flushBatchRequests] sentSynchronously]; // Empty flush

    int i;
    [p opByteSOnewayCallCount];
    for(i = 0 ; i < 30 ; ++i)
    {
        [batch begin_opByteSOneway:bs1];
    }

    int count = 0;
    while(count < 27) // 3 * 9 requests auto-flushed.
    {
        count += [p opByteSOnewayCallCount];
        [NSThread sleepForTimeInterval:10 / 1000.0];
    }

    if([batch ice_getConnection])
    {
        id<TestOperationsMyClassPrx> batch1 = [TestOperationsMyClassPrx uncheckedCast:[p ice_batchOneway]];
        id<TestOperationsMyClassPrx> batch2 = [TestOperationsMyClassPrx uncheckedCast:[p ice_batchOneway]];

        [batch1 end_ice_ping:[batch1 begin_ice_ping]];
        [batch2 end_ice_ping:[batch2 begin_ice_ping]];
        [batch1 end_ice_flushBatchRequests:[batch1 begin_ice_flushBatchRequests]];
        [[batch1 ice_getConnection] close:ICECloseGracefullyAndWait];
        [batch1 end_ice_ping:[batch1 begin_ice_ping]];
        [batch2 end_ice_ping:[batch2 begin_ice_ping]];

        [batch1 ice_getConnection];
        [batch2 ice_getConnection];

        [batch1 end_ice_ping:[batch1 begin_ice_ping]];
        [[batch1 ice_getConnection] close:ICECloseGracefullyAndWait];
        [batch1 end_ice_ping:[batch1 begin_ice_ping]];
        [batch2 end_ice_ping:[batch2 begin_ice_ping]];
    }

    ICEIdentity *identity = [ICEIdentity identity:@"invalid" category:@""];
    id<ICEObjectPrx> batch3 = [batch ice_identity:identity];
    [batch3 begin_ice_ping];
    [batch3 end_ice_flushBatchRequests:[batch3 begin_ice_flushBatchRequests]];

    // Make sure that a bogus batch request doesn't cause troubles to other ones.
    [batch3 begin_ice_ping];
    [batch begin_ice_ping];
    [batch end_ice_flushBatchRequests:[batch begin_ice_flushBatchRequests]];
    [batch begin_ice_ping];
}

File: java-compat/test/src/main/java/test/Ice/operations/BatchOnewaysAMI.java
// **********************************************************************
//
// Copyright (c) 2003-2017 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

package test.Ice.operations;

import java.io.PrintWriter;

import Ice.LocalException;
import test.Ice.operations.Test.Callback_MyClass_opByteSOneway;
import test.Ice.operations.Test.MyClassPrx;
import test.Ice.operations.Test.MyClassPrxHelper;

class BatchOnewaysAMI
{
    private static class Callback
    {
        Callback()
        {
            _called = false;
        }

        public synchronized void check()
        {
            while(!_called)
            {
                try
                {
                    wait();
                }
                catch(InterruptedException ex)
                {
                }
            }

            _called = false;
        }

        public synchronized void called()
        {
            assert (!_called);
            _called = true;
            notify();
        }

        private boolean _called;
    }

    private static void test(boolean b)
    {
        if(!b)
        {
            throw new RuntimeException();
        }
    }

    static void batchOneways(MyClassPrx p, PrintWriter out)
    {
        final byte[] bs1 = new byte[10 * 1024];

        MyClassPrx batch = MyClassPrxHelper.uncheckedCast(p.ice_batchOneway());
        batch.end_ice_flushBatchRequests(batch.begin_ice_flushBatchRequests()); // Empty flush

        test(batch.begin_ice_flushBatchRequests().isCompleted()); // Empty flush
        test(batch.begin_ice_flushBatchRequests().isSent()); // Empty flush
        test(batch.begin_ice_flushBatchRequests().sentSynchronously()); // Empty flush

        for(int i = 0; i < 30; ++i)
        {
            batch.begin_opByteSOneway(bs1, new Callback_MyClass_opByteSOneway()
            {
                @Override
                public void exception(LocalException ex)
                {
                    test(false);
                }

                @Override
                public void response()
                {
                }
            });
        }

        int count = 0;
        while(count < 27) // 3 * 9 requests auto-flushed.
        {
            count += p.opByteSOnewayCallCount();
            try
            {
                Thread.sleep(10);
            }
            catch(InterruptedException ex)
            {
            }
        }

        if(batch.ice_getConnection() != null)
        {
            MyClassPrx batch2 = MyClassPrxHelper.uncheckedCast(p.ice_batchOneway());

            batch.begin_ice_ping();
            batch2.begin_ice_ping();
            batch.end_ice_flushBatchRequests(batch.begin_ice_flushBatchRequests());
            batch.ice_getConnection().close(Ice.ConnectionClose.GracefullyWithWait);
            batch.begin_ice_ping();
            batch2.begin_ice_ping();

            batch.ice_getConnection();
            batch2.ice_getConnection();

            batch.begin_ice_ping();
            batch.ice_getConnection().close(Ice.ConnectionClose.GracefullyWithWait);
            batch.begin_ice_ping().throwLocalException();
            batch2.begin_ice_ping().throwLocalException();
        }

        Ice.Identity identity = new Ice.Identity();
        identity.name = "invalid";
        Ice.ObjectPrx batch3 = batch.ice_identity(identity);
        batch3.begin_ice_ping();
        batch3.end_ice_flushBatchRequests(batch3.begin_ice_flushBatchRequests());

        // Make sure that a bogus batch request doesn't cause troubles to other
        // ones.
        batch3.begin_ice_ping();
        batch.begin_ice_ping();
        batch.end_ice_flushBatchRequests(batch.begin_ice_flushBatchRequests());
        batch.begin_ice_ping();
    }
}
